### DISCUSSION TITLE
how to check if collection has really changed

### DISCUSSION THREAD

**Discussion by unonweb** [payloadcms/payload#3315] [General]
Hi community,
I'd like to check if a collection has really changed before I do filesystem operations.
Does somebody have a recommended approach?
I tried the following but it always returns that the collection has changed...

```js
export default function hasChanged(doc, previousDoc) {
	// att! deletes properties...
	delete previousDoc.__v
	delete previousDoc._id
	delete previousDoc.updatedAt
	delete doc.updatedAt
	return JSON.stringify(doc) !== JSON.stringify(previousDoc)
}
```

**Comment by matthijs166** [comment]
Looks like something I would write to check if 2 objects are different.

Maybe you can also only do it when certain fields change you care about.

**Comment by unonweb** [comment]
Hi Matthijs, thank you for your reply. In the end I did the following:
```js
export default function hasChanged(doc, previousDoc) {
	return !areObjEqual(doc, previousDoc, { ignoreKeys: ['__v', 'updatedAt', '_id'] })
}

function areObjEqual(a, b, { ignoreKeys = [] } = {}) {
	let aMod = Object.entries(a).filter(([key]) => !ignoreKeys.includes(key)).sort()
	let bMod = Object.entries(b).filter(([key]) => !ignoreKeys.includes(key)).sort()
	return JSON.stringify(aMod) == JSON.stringify(bMod)
}
```
This helps me to prevent triggering unnecessary fs operations.

**Reply by matthijs166** [reply]
Converting objects to JSON feels dirty, but it looks clean and if it works it works right? 

Can't you compare only the field values that affect the FS operation?

Also had an idea to maybe hash to object, seems like a JSON stringify is faster:
https://www.measurethat.net/Benchmarks/Show/16519/0/hash-sum-vs-object-hash-vs-json-with-array

But yeah also don't think anyone will notice the performance difference on this level. But yeah, it's satisfying to get right.
